src/Tokenizers/PHP.php

```diff
@@ -945,6 +946,56 @@ protected function tokenize($string)
                 continue;
             }//end if
 
+            /*
+                "readonly" keyword support
+                PHP < 8.1: Converts T_STRING to T_READONLY - Name tokens (T_NAME_*) have to be converted to T_STRING first
+                PHP >= 8.1: Converts some T_READONLY to T_STRING because token_get_all() without the TOKEN_PARSE flag cannot distinguish between them in some situations
+            */
+
+            if ($tokenIsArray === true
+                && (($token[0] === T_STRING && strtolower($token[1]) === 'readonly')
+                || (PHP_VERSION_ID >= 80100 && $token[0] === T_READONLY))
+            ) {
+                // Get the next non-empty token.
+                for ($i = ($stackPtr + 1); $i < $numTokens; $i++) {
+                    if (is_array($tokens[$i]) === false
+                        || isset(Util\Tokens::$emptyTokens[$tokens[$i][0]]) === false
+                    ) {
+                        break;
+                    }
+                }
+
+                $shouldBeReadonly = isset($tokens[$i]) === false || (is_array($tokens[$i]) === true && in_array($tokens[$i][0], [T_STRING, T_PUBLIC, T_PROTECTED, T_PRIVATE]) === true);
```

Comment:
I have a niggly feeling that `T_VAR`, `T_STATIC` and `T_VARIABLE` should also be included in this list.... (and in the tests: `readonly var $property`, `readonly static $property`, `public readonly $property`)

FYI: IIRC, `readonly` did become a full-keyword with only one very specific exception: it's not a keyword when followed by an open parenthesis.

Reply:
https://github.com/squizlabs/PHP_CodeSniffer/pull/3480#discussion_r753674931